internal/directives/json_updater.go

```diff
@@ -0,0 +1,166 @@
+package directives
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+	"github.com/tidwall/sjson"
+	"github.com/xeipuuv/gojsonschema"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+)
+
+func init() {
+	builtins.RegisterPromotionStepRunner(newJSONUpdater(), nil)
+}
+
+// jsonUpdater is an implementation of the PromotionStepRunner interface that
+// updates the values of specified keys in a JSON file.
+type jsonUpdater struct {
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newJSONUpdater returns an implementation of the PromotionStepRunner interface
+// that updates the values of specified keys in a JSON file.
+func newJSONUpdater() PromotionStepRunner {
+	r := &jsonUpdater{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (j *jsonUpdater) Name() string {
+	return "json-update"
+}
+
+// RunPromotionStep implements the PromotionStepRunner interface.
+func (j *jsonUpdater) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	failure := PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}
+
+	if err := j.validate(stepCtx.Config); err != nil {
+		return failure, err
+	}
+
+	// Convert the configuration into a typed struct
+	cfg, err := ConfigToStruct[JSONUpdateConfig](stepCtx.Config)
+	if err != nil {
+		return failure, fmt.Errorf("could not convert config into %s config: %w", j.Name(), err)
+	}
+
+	return j.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates jsonUpdater configuration against a JSON schema.
+func (j *jsonUpdater) validate(cfg Config) error {
+	return validate(j.schemaLoader, gojsonschema.NewGoLoader(cfg), j.Name())
+}
+
+func (j *jsonUpdater) runPromotionStep(
+	_ context.Context,
+	stepCtx *PromotionStepContext,
+	cfg JSONUpdateConfig,
+) (PromotionStepResult, error) {
+	updates := make(map[string]any, len(cfg.Updates))
```

Comment:
What is the point of building this map? You're iterating over the list to build the map, then iterating over the map. Seems the former is unnecessary.

Reply:
> I took inspiration for this from the [yaml-updater](https://github.com/akuity/kargo/blob/main/internal/directives/yaml_updater.go#L69)

I suspected this was the case. 😄 

There are two not-entirely-obvious reasons not to emulate that:

1. `yaml.SetStringsInFile()` wants a map as input. The `yaml-update` step _had_ to do that transformation, whereas here, you don't have any need to do so. You have the luxury of just working with the array you already have in-hand.

2. `yaml.SetStringsInFile()` taking a map as input was misguided in the first place. At the time I wrote it, we were only using it to update _always-pre-existing_ keys in a values.yaml or Chart.yaml file. A map of existing keys to new values served that purpose well, but...

  The more-generic `yaml-update` step was added more recently and I should have accounted better for the possibility of it being used to _add new elements to a list_, and in that case, order matters and order is lost in the transformation to a map.
  
  _I'm actually going to fix this today._

> the map will help in data integrity by deduplicating keys

It won't matter. Since the order of the updates will be preserved, the last one wins -- which is identical behavior to what you get if you iterate over a list and transform it to a map.

Reply:
To perhaps add some clarity...

When there's a map involved, you can't do this:

```
config:
    path: configs/settings.json
    updates:
    - key: list.-1
      value: foo
    - key: list.-1
      value: bar
```

But that should be valid.

You can see I fixed how yaml-update does this in #3156